plonky2-verifier/src/hash/poseidon.cairo

```diff
@@ -0,0 +1,758 @@
+
+    #[test]
+    fn test_sbox_monomial() {
+        let x = gl(0x123456789abcdef0);
+        let y = PoseidonTrait::sbox_monomial(x);
+        assert_eq!(y, gl(11853639751010147186));
+    }
+
+    #[test]
+    fn test_sbox_layer() {
+        let mut state = PoseidonStateArray::default();
+        let mut i: usize = 0;
+
+        loop {
+            if (i >= 12) {
+                break;
+            }
+            state.set(i, gl(i.into()));
+            i += 1;
+        };
+
+        let expected_result = PoseidonStateArray::new(
+            array![
+                gl(0),
+                gl(1),
+                gl(128),
+                gl(2187),
+                gl(16384),
+                gl(78125),
+                gl(279936),
+                gl(823543),
+                gl(2097152),
+                gl(4782969),
+                gl(10000000),
+                gl(19487171)
+            ]
+                .span()
+        );
+
+        PoseidonTrait::sbox_layer(ref state);
+        assert_eq!(state, expected_result);
+    }
+
+    #[test]
+    fn test_mds_row_shf() {
+        let v = array![0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11].span();
+        let r = 2;
+        let res = PoseidonTrait::mds_row_shf(r, v);
+        assert_eq!(res, 1360);
+    }
+
+    #[test]
+    fn test_mds_layer() {
+        let res = PoseidonTrait::mds_layer(
+            PoseidonStateArray::new(
+                array![
+                    gl(0),
+                    gl(1),
+                    gl(128),
+                    gl(2187),
+                    gl(16384),
+                    gl(78125),
+                    gl(279936),
+                    gl(823543),
+                    gl(2097152),
+                    gl(4782969),
+                    gl(10000000),
+                    gl(19487171)
+                ]
+                    .span()
+            )
+        );
+
+        let expected_result = PoseidonStateArray::new(
+            array![
+                gl(914231540),
+                gl(1075416846),
+                gl(855786472),
+                gl(1020513242),
+                gl(547603236),
+                gl(616150402),
+                gl(747427508),
+                gl(449170434),
+                gl(857254888),
+                gl(1108558718),
+                gl(656301516),
+                gl(768889774)
+            ]
+                .span()
+        );
+
+        assert_eq!(res, expected_result);
+    }
+
+    #[test]
+    fn test_full_rounds() {
+        let mut input = PoseidonStateArray::new(
+            array![
+                gl(914231540),
+                gl(1075416846),
+                gl(855786472),
+                gl(1020513242),
+                gl(547603236),
+                gl(616150402),
+                gl(747427508),
+                gl(449170434),
+                gl(857254888),
+                gl(1108558718),
+                gl(656301516),
+                gl(768889774)
+            ]
+                .span()
```

Comment:
Since we're using the previous test's output as the following test's input and because we don't care what these arrays really looks like, we can move all arrays declarations consts which would make it cleaner, not necessary though, this is still okay